### PR TITLE
Reduce copies and allocating strings.

### PR DESCRIPTION
**Pull request by johnou**

### CODE REVIEW

**Comment by johnou** [comment]
@slandelle could I trouble you for a review?

**Comment by johnou** [comment]
@slandelle makes sense, so something like this? (keeping it simple without supporting composite buffers)

```java
  private static byte[] readBytes(ByteBuf byteBuf, int length) {
    if (byteBuf.hasArray()) {
      final int offset = byteBuf.arrayOffset() + byteBuf.readerIndex();
      final byte[] bytes;
      if (offset == 0 && length == byteBuf.array().length) {
        bytes = byteBuf.array();
      } else {
        bytes = Arrays.copyOfRange(byteBuf.array(), offset, offset + length);
      }
      byteBuf.skipBytes(length);
      return bytes;
    }
    byte[] array = new byte[length];
    byteBuf.readBytes(array);
    return array;
  }
```

**Comment by slandelle** [comment]
@johnou Why do you want to get bytes while subject is an UF-8 string?

You can directly create the String from the bytes: `new String(byteBuf.array(), offset, length)`

**Comment by johnou** [comment]
@slandelle oh man I forgot that constructor, ptal.

File: messaging/src/main/java/io/atomix/messaging/impl/MessageDecoder.java [hunk 1]
```diff
@@ -148,6 +185,29 @@ protected void decode(
...
+  static String readString(ByteBuf buffer, int length, Charset charset) {
+    if (buffer.isDirect()) {
+      final String result = buffer.toString(buffer.readerIndex(), length, charset);
+      buffer.skipBytes(length);
+      return result;
+    } else {
+      if (buffer.hasArray()) {
+        final int offset = buffer.arrayOffset() + buffer.readerIndex();
+        final String result;
+        if (offset == 0 && length == buffer.array().length) {
```
Comment (slandelle) [inline messaging/src/main/java/io/atomix/messaging/impl/MessageDecoder.java]: IMHO, testing this is useless and you should just perform the second branch.

File: messaging/src/main/java/io/atomix/messaging/impl/MessageDecoder.java [hunk 2]
```diff
@@ -148,6 +185,29 @@ protected void decode(
...
+          result = new String(buffer.array(), charset);
+        } else {
+          result = new String(buffer.array(), offset, length, charset);
+        }
+        buffer.skipBytes(length);
+        return result;
+      }
+      final byte[] array = new byte[length];
```
Comment (slandelle) [inline messaging/src/main/java/io/atomix/messaging/impl/MessageDecoder.java]: Complicated branches. Why not if(isDirect) elseif (hasArray) else?

File: messaging/src/main/java/io/atomix/messaging/impl/MessageDecoder.java (hunk 1 above)
Comment (johnou) [reply messaging/src/main/java/io/atomix/messaging/impl/MessageDecoder.java]: removed.

File: messaging/src/main/java/io/atomix/messaging/impl/MessageDecoder.java (hunk 2 above)
Comment (johnou) [reply messaging/src/main/java/io/atomix/messaging/impl/MessageDecoder.java]: sgtm, updated.

**Comment by johnou** [comment]
@slandelle thanks for the review!

**Comment by kuujo** [comment]
Why are we replacing `ReplayingDecoder` with custom checkpointing?

**Comment by johnou** [comment]
http://normanmaurer.me/presentations/2014-facebook-eng-netty/slides.html#13.0

More overhead and requires handling ReplayingError.

**Comment by kuujo** [comment]
Nice sounds good

**Pull request merged**